api/v1alpha1/cidrpool_test.go

```diff
@@ -0,0 +1,533 @@
+				To(And(
+					ContainSubstring("spec.perNodeNetworkPrefix"),
+				))
+		})
+		It("Invalid PerNodeNetworkPrefix - smaller than 31 for IPv4 pool", func() {
+			cidrPool := v1alpha1.CIDRPool{
+				ObjectMeta: metav1.ObjectMeta{Name: "test"},
+				Spec: v1alpha1.CIDRPoolSpec{
+					CIDR:                 "192.168.0.0/16",
+					PerNodeNetworkPrefix: 32,
+				}}
+			Expect(cidrPool.Validate().ToAggregate().Error()).
+				To(And(
+					ContainSubstring("spec.perNodeNetworkPrefix"),
+				))
+		})
+		It("Invalid PerNodeNetworkPrefix - smaller than 127 for IPv6 pool", func() {
+			cidrPool := v1alpha1.CIDRPool{
+				ObjectMeta: metav1.ObjectMeta{Name: "test"},
+				Spec: v1alpha1.CIDRPoolSpec{
+					CIDR:                 "2001:db8:3333:4444::0/64",
+					PerNodeNetworkPrefix: 128,
+				}}
+			Expect(cidrPool.Validate().ToAggregate().Error()).
+				To(And(
+					ContainSubstring("spec.perNodeNetworkPrefix"),
+				))
+		})
+		It("Valid PerNodeNetworkPrefix - match CIDR prefix size - ipv4", func() {
+			cidrPool := v1alpha1.CIDRPool{
+				ObjectMeta: metav1.ObjectMeta{Name: "test"},
+				Spec: v1alpha1.CIDRPoolSpec{
+					CIDR:                 "192.168.0.0/16",
+					PerNodeNetworkPrefix: 16,
+				}}
+			Expect(cidrPool.Validate()).To(BeEmpty())
+		})
+		It("Valid PerNodeNetworkPrefix - match CIDR prefix size - ipv6", func() {
+			cidrPool := v1alpha1.CIDRPool{
+				ObjectMeta: metav1.ObjectMeta{Name: "test"},
+				Spec: v1alpha1.CIDRPoolSpec{
+					CIDR:                 "2001:db8:3333:4444::0/64",
+					PerNodeNetworkPrefix: 64,
+				}}
+			Expect(cidrPool.Validate()).To(BeEmpty())
+		})
+		It("Invalid NodeSelector - unknown operator", func() {
+			cidrPool := v1alpha1.CIDRPool{
+				ObjectMeta: metav1.ObjectMeta{Name: "test"},
+				Spec: v1alpha1.CIDRPoolSpec{
+					CIDR:                 "192.168.0.0/16",
+					PerNodeNetworkPrefix: 24,
+					NodeSelector: &corev1.NodeSelector{
+						NodeSelectorTerms: []corev1.NodeSelectorTerm{{
+							MatchExpressions: []corev1.NodeSelectorRequirement{{
+								Key:      "foo.bar",
+								Operator: "unknown",
+							}},
+						}},
+					},
+				},
+			}
+			Expect(cidrPool.Validate().ToAggregate().Error()).
+				To(And(
+					ContainSubstring("spec.nodeSelectorTerms"),
+				))
+		})
+		It("Invalid GatewayIndex - too large", func() {
+			cidrPool := v1alpha1.CIDRPool{
+				ObjectMeta: metav1.ObjectMeta{Name: "test"},
+				Spec: v1alpha1.CIDRPoolSpec{
+					CIDR:                 "192.168.0.0/16",
+					PerNodeNetworkPrefix: 24,
+					GatewayIndex:         255,
+				},
+			}
+			Expect(cidrPool.Validate().ToAggregate().Error()).
+				To(And(
+					ContainSubstring("spec.gatewayIndex"),
+				))
+		})
+		It("Valid GatewayIndex - can use index 1 for /31", func() {
+			cidrPool := v1alpha1.CIDRPool{
+				ObjectMeta: metav1.ObjectMeta{Name: "test"},
+				Spec: v1alpha1.CIDRPoolSpec{
+					CIDR:                 "192.168.0.0/16",
+					PerNodeNetworkPrefix: 31,
+					GatewayIndex:         1,
```

Comment:
I'm not sure if the this suggestion is better, but was wondering if it makes sense to have index 0 and 1 for point to point instead of 1 and 2. In theory, the index 2 is already out of bounds of the subnet.

Reply:
Yeah, I agree that index 1 is a bit confusing for point to point networks. I updated the API and implementation to allow index 0 for point to point networks. GatawayIndex is always an index from 0, but for non point to point networks we will reject index 0, because we should not allow network address (e.g. 192.168.0.0/24) of the prefix as a gateway.